### PR TITLE
Secret sharing

### PR DESCRIPTION
**Pull request by hifi**
WIP, just types

### CODE REVIEW

File: crypto/sql_store_upgrade/12-secrets.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+-- v12: Add crypto_secrets table
```
Comment (tulir) [inline crypto/sql_store_upgrade/12-secrets.sql:1]:
```suggestion
-- v12 (compatible with v9+): Add crypto_secrets table
```

File: crypto/sql_store_upgrade/00-latest-revision.sql [hunk 2]
```diff
@@ -1,4 +1,4 @@
--- v0 -> v11: Latest revision
+-- v0 -> v12: Latest revision
```
Comment (tulir) [inline crypto/sql_store_upgrade/00-latest-revision.sql:1]:
```suggestion
-- v0 -> v12 (compatible with v9+): Latest revision
```

File: crypto/sql_store_upgrade/00-latest-revision.sql [hunk 3]
```diff
@@ -93,3 +93,8 @@ CREATE TABLE IF NOT EXISTS crypto_cross_signing_signatures (
 	signature      CHAR(88) NOT NULL,
 	PRIMARY KEY (signed_user_id, signed_key, signer_user_id, signer_key)
 );
+
+CREATE TABLE IF NOT EXISTS crypto_secrets (
+	name   TEXT PRIMARY KEY NOT NULL,
+	secret bytea NOT NULL
```
Comment (tulir) [inline crypto/sql_store_upgrade/00-latest-revision.sql:99]:
```suggestion
	name   TEXT  PRIMARY KEY NOT NULL,
	secret bytea NOT NULL
```

File: id/crypto.go [hunk 4]
```diff
@@ -153,3 +153,13 @@ type CrossSigningKey struct {
...
 }
+
+// Secret storage keys
+type Secret string
+
+const (
+	SecretXSMaster       Secret = "m.cross_signing.master"
+	SecretXSSelfSigning  Secret = "m.cross_signing.self_signing"
+	SecretXSUserSigning  Secret = "m.cross_signing.user_signing"
+	SecretMegolmBackupV1 Secret = "m.megolm_backup.v1"
```
Comment (tulir) [inline id/crypto.go:164]: These values should probably be deduplicated with the account data type, probably by referencing these constants in the account data types (as in `AccountDataCrossSigningMaster = Type{string(id.SecretXSMaster), AccountDataEventType}`)

**Comment by hifi** [comment]
Quite untested still but it's a base.

**Review by tulir — APPROVED** [review]

**Pull request merged**